internal/report/snapshot/filter_traces.go

```diff
@@ -0,0 +1,121 @@
+
+	// Copy old resource to filtered resource
+	resource := resourceSpan.Resource()
+	resource.CopyTo(filteredResourceSpans.Resource())
+
+	// Apply query to resource
+	queryMatchesResource := true // default to true if no query specified
+	if searchQuery != nil {
+		queryMatchesResource = queryMatchesMap(resource.Attributes(), *searchQuery)
+	}
+
+	scopeSpans := resourceSpan.ScopeSpans()
+	for i := 0; i < scopeSpans.Len(); i++ {
+		filteredScopeSpans := filterScopeSpans(scopeSpans.At(i), queryMatchesResource, searchQuery, minimumTimestamp)
+
+		// Don't append empty scope spans
+		if filteredScopeSpans.Spans().Len() != 0 {
+			filteredScopeSpans.CopyTo(filteredResourceSpans.ScopeSpans().AppendEmpty())
+		}
+	}
+
+	return filteredResourceSpans
+}
+
+// filterScopeSpans filters out spans that do not match the query and minimumTimestamp, returning a new ptrace.ScopeSpans without the filtered spans.
+// queryMatchesResource indicates if the query string matches the resource associated with this ScopeSpans.
+func filterScopeSpans(scopeSpans ptrace.ScopeSpans, queryMatchesResource bool, searchQuery *string, minimumTimestamp *time.Time) ptrace.ScopeSpans {
+	filteredTraceSpans := ptrace.NewScopeSpans()
+	spans := scopeSpans.Spans()
+	for i := 0; i < spans.Len(); i++ {
+		span := spans.At(i)
+		if spanMatches(span, queryMatchesResource, searchQuery, minimumTimestamp) {
+			span.CopyTo(filteredTraceSpans.Spans().AppendEmpty())
+		}
+	}
+
+	return filteredTraceSpans
+}
+
+// spanMatches returns true if the query matches either the resource or span, AND the min timestamp.
+func spanMatches(s ptrace.Span, queryMatchesResource bool, searchQuery *string, minimumTimestamp *time.Time) bool {
+	queryMatchesSpan := true // default to true if no query specified
+	// Skip this check if we already know the query matches the resource
+	if !queryMatchesResource && searchQuery != nil {
+		queryMatchesSpan = spanMatchesQuery(s, *searchQuery)
+	}
+
+	timestampMatches := true // default to true if no timestamp specified
+	if minimumTimestamp != nil {
+		timestampMatches = spanMatchesTimestamp(s, *minimumTimestamp)
+	}
+
+	queryMatches := queryMatchesResource || queryMatchesSpan
+
+	return queryMatches && timestampMatches
+}
+
+// spanMatchesTimestamp determines if the span came after the provided timestamp
+func spanMatchesTimestamp(s ptrace.Span, minTime time.Time) bool {
+	return s.EndTimestamp() >= pcommon.NewTimestampFromTime(minTime)
+}
+
+// spanMatchesQuery determines if the given span matches the given query string
+func spanMatchesQuery(span ptrace.Span, query string) bool {
+	return queryMatchesMap(span.Attributes(), query) ||
+		strings.Contains(span.Name(), query) ||
+		strings.Contains(span.TraceID().String(), query) ||
+		strings.Contains(span.SpanID().String(), query) ||
+		strings.Contains(span.ParentSpanID().String(), query) ||
+		strings.Contains(span.StartTimestamp().String(), query) ||
+		strings.Contains(span.EndTimestamp().String(), query)
```

Comment:
The ticket lists:
```
all attributes, resource attributes, name, event, kinds
```
I think we're missing kinds. I'm not sure if we should actually match events, I don't think we show them in the UI IIRC.

Also, do we want matching on the timestamp? If we want to we should go back in a follow-up PR and add it for logs.

Reply:
Yeah, use your best judgment on this. I'm not as familiar with the trace structure so I wasn't sure what should be checked.

Reply:
Timestamp matching requires parsing the search as a timestamp to be reliable, so we should probably leave it out for now. I also don't know how useful span ID/parent ID will be as these are pretty ephemeral. It seemed like low-hanging fruit, and its absence would frustrate a user who had a good use case (particularly for parent ID in a long-running span)

- [ ]

Reply:
Yeah, I think the span IDs are fine to keep, and unlikely to conflict with other things since they are unique IDs.